### DISCUSSION TITLE
Vips command to find out number of levels in multipage tiff

### DISCUSSION THREAD

**Discussion by urmilas239** [libvips/libvips#2127] [Q&A]
Hi,  How can I find out, using vips command line, the number of levels in a multipage tiff? I am looking for something like what imagemagick's "identify -verbose '%n' imagename.tiff" would return , using vips. Thanks.

**Comment by urmilas239** [comment]
vipsheader --all imagename.tiff
gives the information about imagename.tiff[0]

**Comment by jcupitt** [comment]
Hello, you have to keep asking for pages until you get an error. This ought to be improved.

```
john@kiwi:~/pics$ page=0
john@kiwi:~/pics$ while vipsheader k2.tif[page=$page]; do ((page+=1)); done
k2.tif: 1450x2048 uchar, 3 bands, srgb, tiffload
vipsheader: tiff2vips: TIFF does not contain page 1
john@kiwi:~/pics$ echo $page
1
```

If you know all the pages are the same size, you can just ask for all pages together:

```
john@kiwi:~/pics$ vipsheader x.tif
x.tif: 595x841 uchar, 4 bands, srgb, tiffload
john@kiwi:~/pics$ vipsheader x.tif[n=-1]
x.tif: 595x48778 uchar, 4 bands, srgb, tiffload
```

So the number of pages is 48778 / 841. It'll fail if any pages differ in size though.

**Comment by urmilas239** [comment]
That explains it. Thank you very much for the reply.

**Comment by jakobgager** [comment]
I was just facing the same challenge and discovered that if I use the following

```
vipsheader myimage.tif[n=1000]
```
where 1000 is a number larger than the number of expected pages, I directly get the number of pages via the error message:

```
vipsheader: tiff2vips: TIFF does not contain page 2
```
So this tiff contains 2 pages (page 0 and 1).

I just checked a few tif files and it seams to work reasonable. Is there something which may lead to wrong results?

Thank you for your awesome library!

**Comment by jcupitt** [comment]
Hi @jakobgager, there's a better solution now: you can use the `n-pages` metadata item:

```
$ vipsheader -f n-pages nipguide.pdf 
58
```

You can use it like this in a script:

```bash
file=nipguide.pdf
width=$(vipsheader -f width $file)
n_pages=$(vipsheader -f n-pages $file)
```

There are a lot of useful metadata items, try `vipsheader -a` on an image to see them all. It works for GIF and multi-page TIFF too.

**Comment by jakobgager** [comment]
Great to know! Works perfect! Thanks

Unfortunately my tiff files always produce some warnings like 
`(vipsheader:16868): VIPS-WARNING **: 11:59:50.118: Unknown field with tag 292 (0x124) encountered`, 
so I still have to parse the output.  Can I suppress these warnings somehow?

But this is a different question 😉

**Comment by jcupitt** [comment]
You shouldn't need to parse the output -- the warnings are sent to `stderr` and the number of pages to `stdout`.

For example:

```
$ vipsheader -f n-pages TesterLAB.tif

(vipsheader:360825): VIPS-WARNING **: 13:06:20.716: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360825): VIPS-WARNING **: 13:06:20.718: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360825): VIPS-WARNING **: 13:06:20.718: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360825): VIPS-WARNING **: 13:06:20.718: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360825): VIPS-WARNING **: 13:06:20.718: Incompatible type for "RichTIFFIPTC"; tag ignored
1
```

Looks bad, but the warnings are actually on stderr. You can do:

```
$ n_pages=$(vipsheader -f n-pages TesterLAB.tif)

(vipsheader:360840): VIPS-WARNING **: 13:06:36.211: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360840): VIPS-WARNING **: 13:06:36.213: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360840): VIPS-WARNING **: 13:06:36.213: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360840): VIPS-WARNING **: 13:06:36.213: Incompatible type for "RichTIFFIPTC"; tag ignored

(vipsheader:360840): VIPS-WARNING **: 13:06:36.213: Incompatible type for "RichTIFFIPTC"; tag ignored
$ echo $n_pages
1
```

If you set `VIPS_WARNING` you can hide warnings:

```
$ VIPS_WARNING=1 vipsheader -f n-pages TesterLAB.tif
1
```

**Reply by jakobgager** [reply]
You are totally right I just discovered the different streams (I was just using the wrong tool for this before).
Anyway thanks for the hint with the environment variable!
Have a nice day and best regards!
